File: deal-mgmt/app/schemas/consortium.py
"""컨소시엄/공동투자 매핑 스키마."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.models.enums import ConsortiumStatus


class ConsortiumMappingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    lead_buyer_id: uuid.UUID
    lead_buyer_name: str
    co_investor_buyer_id: uuid.UUID
    co_investor_buyer_name: str
    status: ConsortiumStatus
    equity_share_pct: Decimal | None = None
    notes: str | None = None
    created_at: datetime
    updated_at: datetime

    @field_serializer("equity_share_pct")
    @classmethod
    def _serialize_decimal(cls, v: Decimal | None) -> str | None:
        return str(v) if v is not None else None


class ConsortiumMappingCreate(BaseModel):
    lead_buyer_id: uuid.UUID
    co_investor_buyer_id: uuid.UUID
    status: ConsortiumStatus = ConsortiumStatus.TAPPING
    equity_share_pct: Decimal | None = Field(None, ge=0, le=100)
    notes: str | None = Field(None, max_length=2000)


class ConsortiumMappingUpdate(BaseModel):
    status: ConsortiumStatus | None = None
    equity_share_pct: Decimal | None = Field(None, ge=0, le=100)
    notes: str | None = Field(None, max_length=2000)

File: amic-platform/src/modules/ma/types/consortium.ts
export type ConsortiumStatus = "TAPPING" | "CONFIRMED" | "DROPPED";

export interface ConsortiumMapping {
  id: string;
  transaction_id: string;
  lead_buyer_id: string;
  lead_buyer_name: string;
  co_investor_buyer_id: string;
  co_investor_buyer_name: string;
  status: ConsortiumStatus;
  equity_share_pct: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface ConsortiumMappingCreate {
  lead_buyer_id: string;
  co_investor_buyer_id: string;
  status?: ConsortiumStatus;
  equity_share_pct?: number;
  notes?: string;
}

export interface ConsortiumMappingUpdate {
  status?: ConsortiumStatus;
  equity_share_pct?: number;
  notes?: string;
}
